comment/mixins.py

```diff
@@ -1,113 +1,103 @@
+    def has_object_permission(self, request, obj):
+        return request.user == obj.user
+
+    def dispatch(self, request, *args, **kwargs):
+        obj = self.get_object()
+        if not self.has_object_permission(request, obj):
+            return self.handle_no_permission()
+        return super().dispatch(request, *args, **kwargs)
+
+
+class CanDeleteMixin(ObjectLevelMixin, ValidatorMixin, abc.ABC):
+    def has_object_permission(self, request, obj):
+        return request.user == obj.user or is_comment_admin(request.user) \
+                or (obj.is_flagged and is_comment_moderator(request.user))
+
+    def dispatch(self, request, *args, **kwargs):
+        obj = self.get_object()
+        if not self.has_object_permission(request, obj):
+            return self.handle_no_permission()
+        return super().dispatch(request, *args, **kwargs)
+
+
+class BaseFlagMixin(ObjectLevelMixin, abc.ABC):
+    def dispatch(self, request, *args, **kwargs):
+        if not getattr(settings, 'COMMENT_FLAGS_ALLOWED', 0):
+            return HttpResponseForbidden(_('Flagging system must be enabled'))
         return super().dispatch(request, *args, **kwargs)
 
 
-class ContentTypeMixin:
-    """
-    When using this with DRF intialize `api=True` in the class.
-    Used for provinding validation of query parameters in request to match
-    a valid ContentType. Upon successful validation, `model_name`, `model_id`, `app_name`
-    can be accessed directly as class attributes.
-    """
-    api = False
-    error = ''
-
-    def _set_model_name_or_error(self, val):
-        if not val:
-            self.error = "model name must be provided"
-            return
-        self.model_name = val
-
-    def _set_model_id_or_error(self, val):
-        if not val:
-            self.error = "model id must be provided"
-            return
-        self.model_id = val
-
-    def _set_app_name_or_error(self, val):
-        if not val:
-            self.error = 'app name must be provided'
-            return
-
-        if not ContentType.objects.filter(app_label=val).exists():
-            self.error = f'{val} is NOT a valid app name'
-            return
-        self.app_name = val
-
-    def _set_error(self):
-        try:
-            model_name = self.model_name.lower()
-            ct = ContentType.objects.get(model=model_name).model_class()
-            model_class = ct.objects.filter(id=self.model_id)
-            if not model_class.exists() and model_class.count() != 1:
-                self.error = f"{self.model_id} is NOT a valid model id for the model {self.model_name}"
-        except ContentType.DoesNotExist:
-            self.error = f"{self.model_name} is NOT a valid model name"
-        except ValueError:
-            self.error = f"model id must be an integer, {self.model_id} is NOT"
+class CanSetFlagMixin(BaseFlagMixin, abc.ABC):
+    def has_object_permission(self, request, obj):
+        """user cannot flag their own comment"""
```

Comment:
any particular reason why? This seems a bit opiniated to me.

Suppose I comment something that seems funny to me today but return 2 months later to become more sensitive and find it may be nasty towards a community. I would want to flag my own comment for the moderators to decide if the comment should be flagged or not is what I think.

But this is petty thing, feel free to ignore

Reply:
Yes. This doesn't make sense.